app/controllers/notifications/NotificationCallbackController.scala

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.notifications
+
+import config.AppConfig
+import javax.inject.{Inject, Singleton}
+import play.api.Logger
+import play.api.mvc.{Action, ControllerComponents}
+import services.NotificationService
+import uk.gov.hmrc.http.BadRequestException
+import uk.gov.hmrc.play.bootstrap.controller.BackendController
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.xml.NodeSeq
+
+@Singleton
+class NotificationCallbackController @Inject()(notificationsService: NotificationService, cc: ControllerComponents)(implicit ec: ExecutionContext)
+    extends BackendController(cc) {
+
+  private val logger = Logger(this.getClass)
+
+  def onNotify: Action[NodeSeq] = Action.async(parse.xml) { implicit req =>
+    val notification = req.body
+
+    notificationsService.save(notification).map {
+      case Right(_) => Accepted
+      case Left(e: BadRequestException) =>
+        logger.warn(s"Failed to save invalid notification: $notification", e)
+        Accepted
```

Comment:
Is that right?  How does the caller get informed that the notification was invalid?

Reply:
That's a good point. This is from old implementation and right now this case can be removed.

Background why this looks like this:
This case covered the situation when we received the notification without all information inside. In that case when we didn't return the Accepted customs-notification will send this notification over and over again. In cases that the service retrieve the notification and notification has missing some fields etc we must return the Accepted and investigate the situation. In other case we will have the wrong notification coming to us every few seconds (or other time period) until timeout (14 days was before, not sure how it looks like right now in customs notifications).

There is a golden rule for notifications, always return the Accepted. The same approach is in different our projects. The only case that we would like to return something different that Accepted is when we have a problem with mongo and we can't save the notification. In that case the problem can be temporary so we would like to receive this notification once again.

Reply:
It's good to follow the "golden rules"!   So basically, we have to respond "accepted" even if we get sent rubbish otherwise we get spammed.   Nice.